javalib/src/main/scala/java/net/InetAddress.scala

```diff
@@ -88,8 +90,8 @@ class InetAddress protected (ipAddress: Array[Byte], originalHost: String)
     if (ipAddress.length == 4) {
       formatIn4Addr(bytes)
     } else if (ipAddress.length == 16) {
-      if (isIPv4MappedAddress(bytes)) {
-        formatIn4Addr(extractIP4Bytes(bytes).at(0))
+      if (SocketHelpers.isIPv4MappedAddress(bytes)) {
+        formatIn4Addr(ipAddress.at(12))
```

Comment:
I'm not sure if this should be done. We should either
1. handle IPv4 addresses mapped to IPv6 as `Inet4Address` address directly
2. Have this check in many more places. eg. `isLoopbackAddress` with `127.0.0.1` mapped into an `Inet6Address` returns `false` because we only check `::1`.

I'm really in favor of 1. since address comparison gets difficult when comparing `Inet4Address` and IPv4 mapped into an `Inet6Address`

Reply:
This is complicated.   Sorry to be terse, I am running on stolen time now. Forgive
any harshness or discouragement.

I will re-read this when I have more time.  Right now I will give first impressions. 

A) What is the presenting needing to be solved?  Pardon me if it is evident and I am missing it.
B) 
```
if (SocketHelpers.isIPv4MappedAddress(bytes)) {
        formatIn4Addr(ipAddress.at(12))
}
```
  IIRC, the formatIn4Addr() is done here because that is the way that the the
  Java 8 VM formats such an address. Seemed odd to me also, but SN tries to
  be bug-for-bug compatible with JVM.
    
C) re: `handle IPv4 addresses mapped to IPv6 as Inet4Address address directly`
     The two are conceptually distinct and should not be conflated.

D)  I can not longer quote chapter & verse, but the one true IPv6 loopback address is
      (the full expansion of) "::1".  A mapped 127.0.0.x is most definitely _not_ an IPv6 
      loopback.  Try it on JVM (scalacli is useful). I think/hope you will get the same results.

Reply:
A) I faced an issue during implementation where I created IPv4 mapped to Ipv6 instead of creating an IPv4 address. Filtering out packets when socket was connected was dropping everything.

B) It  doesn't look to be matching the JVM behaviour
```
scala> val ip = java.net.Inet6Address.getByAddress(
  "", 
  Array.fill[Byte](8)(0) ++ Array.fill[Byte](4)(127) ++ Array[Byte](127, 0, 0, 1),
  -1
)
val ip: java.net.Inet6Address = /0:0:0:0:0:0:7f00:1
```
All IPv6 addresses are printed following the IPv6 notation. Mapped address are not converted to IPv4 format.

C) Right, this is actually fine. Addresses are most likely created from `InetAdress.getByAddress` that unwraps mapped IPv4 addresses. When manually forcing wrapping, equality with the non wrapped is lost.

D) Correct too.  `127.0.0.x ` is not an IPv6 loopback

Reply:
re: All IPv6 addresses are printed following the IPv6 notation. Mapped address are not converted to IPv4 format.

That seems reasonable & believable.  I will have to study why the wart in question exists and if it is correct
in its context.  Warts are _expensive.


There may be a transcription error in the "val ip:" line here.
When I run the code above in scala-cli (JVM) I get the 16 bytes what I would expect:
`val ip: java.net.Inet6Address = /0:0:0:0:7f7f:7f7f:7f00:1`

Reply:
A dive into the land of details from which most people run, but which are necessary for a 
correct implementation.

re: When manually forcing wrapping, equality with the non wrapped is lost

It is not that equality is lost but rather that it was never there.   The two "have equality" in
the human sense, but not in the sense defined by the relevant IETF RFCs (Internet Engineering 
Task Force, Request for Comments).  That latter sense is what needs to be operative here.

Relying upon a fine points the two "correspond" but have different types.  One would have
to check the IPv6 RFC where they very carefully described this.

The of course, the passage of time may have made me completely wrong.